[16.0][Add] sale_order_blanket_order

lmignon opened the pull request:
New target repo for https://github.com/OCA/sale-workflow/pull/3436

* [x] depends on https://github.com/OCA/sale-workflow/pull/3423
This module extends the functionality of Sale Order to support Blanket
Order and Call-off Order.

Blanket Order
=============

A Blanket Order is a standard sales order with the following specific
features:

- Type: Classified as "Blanket Order".
- Defined Duration: Includes a validity period (end date).
- Payment Terms: Allows selection of preferred terms (e.g., 90 days end
  of month, upon delivery, etc.).
- Invoicing Policy: Can be based on product settings or the order
  itself.
- Stock Reservation: Allows advance reservation of sold quantities.
- Handling Unfulfilled Quantities: Provides options for dealing with
  undelivered quantities upon order expiration.
- Prices are calculated based on existing rules since it is a standard
  sales order type.

The blanket order serves as the central element triggering stock
management and invoicing mechanisms.

Stock Management
----------------

Delivered quantities are tracked on the sales order lines as with
regular sales orders. By default, the stock is not reserved upon
confirmation of the blanket order. This is achieved by using the OCA
module
`sale_manual_delivery <https://pypi.org/project/odoo-addon-sale-manual-delivery/>`__.
As a result, the stock will be reserved only when a call-off order is
created for the quantity to be delivered.

In some cases, you may want to reserve stock upon confirmation of the
blanket order. This can be achieved by using the OCA module
`sale_order_blanket_order_stock_prebook <https://pypi.org/project/odoo-addon-sale-order-blanket-order-stock-prebook/>`__.
This module extends the functionality of Sale Blanket Order to support
the reservation of stock for future consumption by call-off orders. The
reservation is done at the time of the blanket order confirmation for a
consumption starting at the validity start date of the blanket order.
This behavior can be configured on the blanket order.

Invoicing
---------

Standard invoicing policies apply (e.g., invoice on order or on
delivery). Payment terms are configurable per order. Prepayment can be
enforced by configuring the invoicing policy at the order level using
the OCA module
`sale_invoice_policy <https://pypi.org/project/odoo-addon-sale-invoice-policy/>`__.

Consumption Management
----------------------

A wizard will be available on the blanket order to initiate a delivery.
It allows users to select products and quantities for delivery. This
action creates a Call-off Order linked to the blanket order.

Call-off Order
==============

A Call-off Order is a standard sales order with these specific
characteristics:

- Type: Classified as "Call-off Order".
- Linked to Blanket Order: Only includes products from the blanket
  order.
- Delivery Release: Enables the release of reserved stock for delivery.
- No Invoicing or Stock Management: These are handled via the linked
  blanket order.

Stock Management
----------------

No delivery is generated directly from the call-off order.

It triggers:

- Release of the reserved quantity in the blanket order.
- Adjustment of stock reservations for the remaining quantities.

Standard Sales Orders
=====================

To support existing workflows (e.g., e-commerce), call-off orders can be
generated transparently from standard sales orders based on product and
availability:

Entire orders may be converted into call-off orders if all products are
linked to a blanket order. Mixed orders split call-off items into a new
call-off order, with both confirmed within the available quantities of
the blanket order.

**Table of contents**

.. contents::
   :local:

Use Cases / Context
===================

When a company sells the same products to the same customers on a
regular basis, it's a common business practice to create a blanket order
that defines the terms and conditions of the sales.

If you need a way to define:

- the terms and conditions of the sales,
- the payment terms,
- the delivery terms,

and also secure the quantities of the products to be delivered, the sale
order blanket order module is the right choice.

This module introduces 2 new kinds of sales orders:

1. Blanket Order: This is a sales order that defines the terms and
   conditions of the sales, the price, the payment terms, the delivery
   terms, and secures the quantities of the products to be delivered.

2. Call of order: This is a sales order linked to a blanket order that
   is created to trigger the delivery of quantities of the products
   secured in the blanket order.

Others modules can be used to provide the same kind of features. For
example, the module
(sale_blanket_order)[`https://pypi.org/project/odoo-addon-sale-blanket-order] <https://pypi.org/project/odoo-addon-sale-blanket-order]>`__
also defines the concept of sale blanket order. The main difference
between the two modules is that the sale order blanket order module
extends the sale order model to add the sale blanket order and the call
off order. This allows to keep the benefits of all the extensions made
on the sale order model by other modules without having to adapt them to
the sale blanket order model (discount, invoicing; inventory process,
...).

Usage
=====

By default, the automatic creation of call-off orders from normal sale
orders containing products part of a blanket order is disabled. To
enable this feature, you need to go into the sales settings and enable
the option "Create Call-Off from SO if possible".

lmignon committed:
[ADD] sale_order_blanket_order, sale_order_blanket_order_stock_prebook

These modules extends the functionality of Sale Order to support Blanket Order and Call-off Order.

lmignon commented:
ping @jbaudoux @rousseldenis

jbaudoux reviewed (approved):
Thanks for this amazing work

lmignon commented:
> Thanks for this amazing work

Can you also take a look at https://github.com/OCA/sale-workflow/pull/3423 ? Thank you

sbejaoui committed:
[IMP] sale_order_blanket_order: allow blanket order overlap for specific products

by default, blanket order overlap is not allowed, ensuring that only one active blanket order applies at a time.
however, for certain products like delivery services, overlapping blanket orders are normal and do not pose an issue.

this commit adds a bool field, allow_blanket_order_overlap, to the product model, allowing specific products to permit
overlapping blanket orders when enabled.

sbejaoui committed:
[IMP] sale_order_blanket_order: prevent call-off with only delivery products

At the EOL of a blanket order, if the only remaining product to deliver is a
delivery product and the user hasn't included it in any of the call-off orders,
it makes no sense to create a call-off order containing only a delivery product
without tangible items.
Such orders would serve no practical purpose and could lead to confusion.
Ensuring that call-off orders include at least one actual product maintains
meaningful and logical orders.

sbejaoui committed:
[FIX] _compute_qty_to_deliver depends on order_type

lmignon commented:
/ocabot merge nobump
